### PR TITLE
hub: Add merchant name to notifications

### PR DESCRIPTION
**Pull request by backspace**
This follows #2434 to parse/fetch DIDs and insert the merchant name in the notification:

image

It ignores failures that might arise when parsing the DID and parsing and fetching its associated JSON and just produces a generic “your business” notification.

I extracted the `JSONAPIDocument` type that was scattered through various files.

### CODE REVIEW

**Review by lukemelia — APPROVED** [review]
I wonder if we will need/want to cache this in memory.

**Comment by backspace** [comment]
yes, we might want to add caching here as we did with “off chain JSON” in web client

File: packages/hub/tasks/notify-customer-payment.ts [hunk 1]
```diff
@@ -65,9 +69,23 @@ export default class NotifyCustomerPayment {
...
+    try {
+      if (result.merchantSafe?.infoDid) {
+        let merchantInfo = await this.merchantInfo.getMerchantInfo(result.merchantSafe.infoDid);
+
+        if (merchantInfo) {
+          merchantName = ` ${merchantInfo.name}`;
+        }
+      }
+    } catch (e) {
+      // Silently ignore and just show notification without merchant name
```
Comment (Aierie) [inline packages/hub/tasks/notify-customer-payment.ts:83]: Is it worth logging this to Sentry?

File: packages/hub/tasks/notify-merchant-claim.ts [hunk 2]
```diff
@@ -53,10 +57,27 @@ export default class NotifyMerchantClaim {
...
+    }
+
     let token = result.token.symbol;
     let notifiedAddress = result.merchantSafe.merchant.id;
     let amountInWei = result.amount;
-    let message = `You just claimed ${Web3.utils.fromWei(amountInWei)} ${token} from your business account`;
+
+    let message = `You just claimed ${Web3.utils.fromWei(
+      amountInWei
+    )} ${token} from your${merchantName} business account`;
```
Comment (Aierie) [inline packages/hub/tasks/notify-merchant-claim.ts:80]: I think this reads better than `Revenue deposited into account for [Merchant name].` in CS-2617. Will ask in the issue.

File: packages/hub/tasks/notify-customer-payment.ts [hunk 3]
```diff
@@ -65,9 +69,23 @@ export default class NotifyCustomerPayment {
...
+    }
+
     let notifiedAddress = result.merchant.id;
     let spendAmount = result.spendAmount;
-    let message = `Your business received a payment of §${spendAmount}`;
+    let message = `Your business${merchantName} received a payment of §${spendAmount}`;
```
Comment (Aierie) [inline packages/hub/tasks/notify-customer-payment.ts:88]: Will also ask about this v `[Merchant name] received a new payment.` in CS-2616.

File: packages/hub/node-tests/services/merchant-info-test.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+  it('returns a MerchantInfo object extracted from a fetched document', async function () {
+    let mockMerchantInfoDocument = {
+      meta: { network: 'sokol' },
+      data: {
+        id: '05074049-be24-45fe-81ba-8e3c1a623aa7',
+        type: 'merchant-infos',
+        attributes: {
+          did: 'did:cardstack:1m1C1LK4xoVSyybjNRcLB4APbc07954765987f62',
+          name: 'a long long long long multi-word name that will cause problems because it is so long it really maybe shouldn’t even be permitted amirite maybe there should be (more of?) a limit',
```
Comment (Aierie) [inline packages/hub/node-tests/services/merchant-info-test.ts:16]: yes

File: packages/hub/tasks/notify-customer-payment.ts [hunk 5]
```diff
@@ -65,9 +69,23 @@ export default class NotifyCustomerPayment {
       );
     }
 
+    let merchantName = '';
+
+    try {
+      if (result.merchantSafe?.infoDid) {
+        let merchantInfo = await this.merchantInfo.getMerchantInfo(result.merchantSafe.infoDid);
+
+        if (merchantInfo) {
```
Comment (Aierie) [inline packages/hub/tasks/notify-customer-payment.ts:78]: Maybe worth changing to `merchantInfo?.name`?

File: packages/hub/tasks/notify-merchant-claim.ts [hunk 6]
```diff
@@ -53,10 +57,27 @@ export default class NotifyMerchantClaim {
       throw new Error(`Subgraph did not return information for merchant claim with transaction hash: "${payload}"`);
     }
 
+    let merchantName = '';
+
+    try {
+      if (result.merchantSafe.infoDid) {
+        let merchantInfo = await this.merchantInfo.getMerchantInfo(result.merchantSafe.infoDid);
+
+        if (merchantInfo) {
```
Comment (Aierie) [inline packages/hub/tasks/notify-merchant-claim.ts:66]: Maybe worth changing to `merchantInfo?.name`?

**Review by Aierie — APPROVED** [review]
I think this looks good and it works for me locally.

I wonder if we should be introducing typescript-friendly utilities (types + (de)serialization) for JSON-API?

File: packages/hub/tasks/notify-customer-payment.ts (hunk 1 above)
Comment (backspace) [reply packages/hub/tasks/notify-customer-payment.ts:83]: Maybe? I’m wary of more noisy logging, but perhaps an error here is worth knowing about. I’ll do the same in the other task, thanks.

Comment (backspace) [reply packages/hub/tasks/notify-customer-payment.ts]: done in c241690 🤓

File: packages/hub/tasks/notify-customer-payment.ts (hunk 5 above)
Comment (backspace) [reply packages/hub/tasks/notify-customer-payment.ts]: Agreed, done in c91e5682a, thanks

File: packages/hub/tasks/notify-customer-payment.ts (hunk 3 above)
Comment (backspace) [reply packages/hub/tasks/notify-customer-payment.ts]: Thanks for checking in, I was not thinking too much about wording pending concrete answers, but I’ve updated in c281ddd80 with the decision there.

**Comment by backspace** [comment]
re Typescript and JSON:API, maybe, probably? haha

**Pull request merged**